### ISSUE TITLE
[WEB-967] Issue using cookies with next/headers with Turbopack enabled

### ISSUE THREAD

**Issue by rmsheppard** [vercel/next.js#45371]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: win32
      Arch: x64
      Version: Windows 10 Pro
    Binaries:
      Node: 16.14.2
      npm: N/A
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.1.6-canary.1, 13.1.5
      eslint-config-next: 13.1.5
      react: 18.2.0
      react-dom: 18.2.0
      
    I've tried **Node** version 18 as well as Next both next versions listed above.

### Which area(s) of Next.js are affected? (leave empty if unsure)

Turbopack (--turbo)

### Link to the code that reproduces this issue

n/a

### To Reproduce

```js
/*
The following code will reproduce the issue in any project with turbopack enabled
*/

import React from 'react';
import { cookies } from 'next/headers';
import { redirect } from 'next/navigation';

const Home = () => {

    const nextCookies = cookies();
    const auth = nextCookies.get('ccauth');

    if (!auth) {
    redirect('/login');
  }

    return (
        <div>page</div>
    );
};

export default Home;
```

### Describe the Bug

App breaks when trying to access next/header cookie info on server component

**Unhandled Runtime Error**
1 of 1 
Error: Invariant: Method expects to have **requestAsyncStorage**, none available

### Expected Behavior

Access to header/cookie info is returned

### Which browser are you using? (if relevant)

Chrome Version 109.0.5414.120 (Official Build) (64-bit)

### How are you deploying your application? (if relevant)

n/a

<sub>[WEB-967](https://linear.app/vercel/issue/NEXT-700/issue-using-cookies-with-nextheaders-with-turbopack-enabled)</sub>

**Comment by AndreasWintherMoen** [comment]
Same issue with headers.

**Comment by orels1** [comment]
Same issue here, tried to use `generateStaticParams` with `Supabase`, but got a build error, so I do not seem to be able to utilize static params generation due to this bug.

Requests to supabase in server components overall seem to work though

**Comment by orels1** [comment]
For my specific case I eneded up doing this (since my data doesn't need to be authorized), until the issue is fixed:

```ts
import { headers, cookies } from 'next/headers'
import { RequestCookies } from 'next/dist/server/web/spec-extension/cookies'
import { createServerComponentSupabaseClient } from '@supabase/auth-helpers-nextjs'

import type { Database } from '../lib/database.types'

// This is currently short-circuitet per 
export const createClient = () =>
  createServerComponentSupabaseClient<Database>({
    headers: () => new Headers(),
    cookies: () => new RequestCookies(new Headers()),
  })
  
```

**Comment by WilderDev** [comment]
Any reason why it's happening?

**Comment by leopedroso45** [comment]
same here...

**Comment by orels1** [comment]
Just wanted to comment that even without using turbopack - this is still an issue.

![image](https://user-images.githubusercontent.com/3798928/222252062-9b17c6b8-242a-4351-8c97-59b3046b75bf.png)

**Comment by BierDav** [comment]
Same error here. Are there any plans, fixing this issue?

**Comment by FrimJo** [comment]
I'm facing the same problem where I'm using `react-hook-form` and the `onSubmit` prop on `<form>`.

```ts
// page.tsx

"use client"
import { createUser } from "./actions";

export default function CreateUser() {
    const { register, handleSubmit } = useForm();
    const [ startTransition ] = useTransition();

    return (
        <form onSubmit={handleSubmit((data) => startTransition(() => createUser(data)))}>
            ...
        </form>
    )
}
```

```ts
// actions.ts

"use server"
import { currentUser } from "@clerk/nextjs";
import { CreateUserFormData } from "@/lib/create-user-form";

export async function createUser(data: CreateUserFormData) {
    const clerkUser = await currentUser();
    ....
}
```

```
Error: Clerk: auth() and currentUser() are only supported in App Router (/app directory).
If you're using /pages, try getAuth() instead.
Original error: Error: Invariant: Method expects to have requestAsyncStorage, none available
```

**Comment by Neutronas** [comment]
same issue. Any workarounds so far?

**Comment by 0xcadams** [comment]
This was fixed for me in 13.4.7, previous versions were not working.

**Comment by thorwebdev** [comment]
I'm facing this issue also when working with cookies in server components. Interestingly though locally, everything is working fine (even though there are the error logs), however once deployed to Vercel, database communication isn't working.

For reproduction can use https://github.com/supabase-community/vercel-ai-chatbot or 
[![Deploy with Vercel](https://vercel.com/button)](https://vercel.com/new/clone?repository-url=https%3A%2F%2Fgithub.com%2Fsupabase-community%2Fvercel-ai-chatbot&env=OPENAI_API_KEY&envDescription=You%20must%20first%20activate%20a%20Billing%20Account%20here%3A%20https%3A%2F%2Fplatform.openai.com%2Faccount%2Fbilling%2Foverview&envLink=https%3A%2F%2Fplatform.openai.com%2Faccount%2Fapi-keys&project-name=vercel-ai-chatbot-with-supabase&repository-name=vercel-ai-chatbot-with-supabase&integration-ids=oac_VqOgBHqhEoFTPzGkPd7L0iH6&external-id=https%3A%2F%2Fgithub.com%2Fsupabase-community%2Fvercel-ai-chatbot%2Ftree%2Fmain)

When running the app locally, while the error is printed to the terminal, but information is correctly read from the cookies and JWTs are forwarded to the DB, making it possible to save and load chat history as intended.

Once deployed Vercel, this stops working. You can inspect the logs for the `api/chat` egde function where the error is emitted and the cookie cannot be read. 
<img width="1024" alt="image" src="https://github.com/vercel/next.js/assets/5748289/edb5af84-f949-45d9-920d-3b24631b7686">

Note: I've even implemented the server-only safe guards as recommended here: https://nextjs.org/docs/getting-started/react-essentials#the-server-only-package so I think everything is implemented as intended by Next.js and this seems to be an actual issue.

**Comment by Gbuomprisco** [comment]
I'm hitting the same issue - I confirm it gets fixed by replacing runtime to `nodejs` (without Turbopack)

**Comment by atleugim** [comment]
I'm using `createRouteHandlerClient` from the [@supabase/auth-helpers-nextjs](https://github.com/supabase/auth-helpers) package.

It was throwing errors even when I was trying to login:
_`Method expects to have requestAsyncStorage, none available at Object.cookies ...`_

I was able to fix this by adding the [`import "server-only"`](https://nextjs.org/docs/getting-started/react-essentials#the-server-only-package) in my route-handler without using the `export const runtime = 'nodejs'` flag, now it works fine in ▲ Vercel.

**Comment by ExLineP** [comment]
Getting the same error but switching to nodejs runtime fixes it

**Comment by burkayanduv** [comment]
I have the same issue for Vercel when using `edge` runtime for the `supabase` functions `createServerComponentClient` `createRouteHandlerClient` when `cookies` are used from `next/headers`. Moving to `nodejs` runtime fixes the issue. Waiting for a fix for `edge` runtime.

**Comment by emrecoban** [comment]
> I have the same issue for Vercel when using `edge` runtime for the `supabase` functions `createServerComponentClient` `createRouteHandlerClient` when `cookies` are used from `next/headers`. Moving to `nodejs` runtime fixes the issue. Waiting for a fix for `edge` runtime.

Could you share which pages you added `nodejs` runtime variable? If I add `nodejs` runtime, it throws that error:
```bash
Please make sure that all your non-static routes export the following edge runtime route segment config:
export const runtime = 'edge';
```  

If I add `edge` runtime, it throws another error:
```bash
Method expects to have requestAsyncStorage, none available
```

**Comment by burkayanduv** [comment]
I had the `export const runtime = 'edge';` on the files:

`/app/auth/callback/route.ts`
`/app/error.tsx`
`/app/layout.tsx`
`/app/not-found.tsx`

And the `supabase` functions had `Method expects to have requestAsyncStorage, none available` erros.

Removing all `export const runtime = 'edge';` lines for these files (defaulting back to `nodejs`) fixed the error temporarirly. Waiting for the fix for `edge` runtime now.

**Comment by burkayanduv** [comment]
> > I have the same issue for Vercel when using `edge` runtime for the `supabase` functions `createServerComponentClient` `createRouteHandlerClient` when `cookies` are used from `next/headers`. Moving to `nodejs` runtime fixes the issue. Waiting for a fix for `edge` runtime.
> 
> Could you share which pages you added `nodejs` runtime variable? If I add `nodejs` runtime, it throws that error:
> 
> ```shell
> Please make sure that all your non-static routes export the following edge runtime route segment config:
> export const runtime = 'edge';
> ```
> 
> If I add `edge` runtime, it throws another error:
> 
> ```shell
> Method expects to have requestAsyncStorage, none available
> ```


I had the export const runtime = 'edge'; on the files:

/app/auth/callback/route.ts
/app/error.tsx
/app/layout.tsx
/app/not-found.tsx

And the supabase functions had Method expects to have requestAsyncStorage, none available erros.

Removing all export const runtime = 'edge'; lines for these files (defaulting back to nodejs) fixed the error temporarirly. Waiting for the fix for edge runtime now.

**Comment by thorwebdev** [comment]
Okay, I think I finally figured this out (at least for the Supabase side of things): It's very important at what point you call the `cookies` method for things to work correctly. 

So rather than handing the `cookies` method to the supabase client, we need to extract the value first, and then hand the supabase client a function that retrieves that value, e.g.:

```js
'use server'
import { createServerActionClient } from '@supabase/auth-helpers-nextjs'
import { cookies } from 'next/headers'
import { Database } from '@/lib/db_types'

const readOnlyRequestCookies = cookies()
const supabase = createServerActionClient<Database>({
  cookies: () => readOnlyRequestCookies
})
```

You can see my PR that fixed this here: https://github.com/supabase-community/vercel-ai-chatbot/pull/8

I do think the issue should probably stay open, because I'm not sure the above workaround should be needed, but yah, not sure what exactly is breaking things when handing the method down.

**Comment by emrecoban** [comment]
@thorwebdev I tried it, but it didn't work. It throws the same error when building application, not after built. Also logging and logout doesn't work properly when it is as you shared.

```javascript
'use server'

import { cookies } from 'next/headers';
import { cache } from 'react';
import { createServerComponentClient } from '@supabase/auth-helpers-nextjs';

const cookieStore = cookies()
export const createServerSupabaseClient = cache(() =>
    createServerComponentClient({ cookies: () => cookieStore })
)

export async function getSession() {
    const supabase = createServerSupabaseClient()
    try {
        const {
            data: { session }
        } = await supabase.auth.getSession()
        return session
    } catch (error) {
        console.error('Error:', error)
        return null
    }
}
```

```bash
13:30:59.279 | ▲  Error: Invariant: Method expects to have requestAsyncStorage, none available
13:30:59.284 | ▲  at h (/opt/buildhome/repo/.next/server/app/(auth)/forgot-password/page.js:46:30261)
13:30:59.285 | ▲  at 79040 (/opt/buildhome/repo/.next/server/app/(auth)/forgot-password/page.js:46:48718)
13:30:59.285 | ▲  at __webpack_require__ (/opt/buildhome/repo/.next/server/edge-runtime-webpack.js:25:43)
13:30:59.285 | ▲  at 91106 (/opt/buildhome/repo/.next/server/app/(auth)/forgot-password/page.js:46:48349)
13:30:59.285 | ▲  at Function.__webpack_require__ (/opt/buildhome/repo/.next/server/edge-runtime-webpack.js:25:43)
13:30:59.286 | ▲  at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
13:30:59.286 | ▲  > Build error occurred
13:30:59.286 | ▲  Error: Failed to collect page data for /forgot-password
13:30:59.287 | ▲    at /opt/buildhome/repo/node_modules/.pnpm/next@13.4.12_@babel+core@7.22.9_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/build/utils.js:1156:15
13:30:59.288 | ▲    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
```

**Comment by thorwebdev** [comment]
cookies() needs to be called inside your getSession method I think.

**Comment by emrecoban** [comment]
@thorwebdev unfortunately the same.

```javascript
'use server'
import { cookies } from 'next/headers';
import { cache } from 'react';
import { createServerComponentClient } from '@supabase/auth-helpers-nextjs';

const cookieStore = cookies()
export const createServerSupabaseClient = cache(() =>
    createServerComponentClient({ cookies: () => cookieStore })
)

export async function getSession() {
    const cookieStoreTwo = cookies()
    const supabase = createServerComponentClient({ cookies: () => cookieStoreTwo })
    try {
        const {
            data: { session }
        } = await supabase.auth.getSession()
        return session
    } catch (error) {
        console.error('Error:', error)
        return null
    }
}
```

```bash
14:05:32.299 | ▲  Error: Invariant: Method expects to have requestAsyncStorage, none available
14:05:32.300 | ▲  at h (/opt/buildhome/repo/.next/server/app/(auth)/register/page.js:46:30261)
14:05:32.300 | ▲  at 79040 (/opt/buildhome/repo/.next/server/app/(auth)/register/page.js:46:48791)
14:05:32.300 | ▲  at __webpack_require__ (/opt/buildhome/repo/.next/server/edge-runtime-webpack.js:25:43)
14:05:32.300 | ▲  at 91106 (/opt/buildhome/repo/.next/server/app/(auth)/register/page.js:46:48351)
14:05:32.301 | ▲  at Function.__webpack_require__ (/opt/buildhome/repo/.next/server/edge-runtime-webpack.js:25:43)
14:05:32.301 | ▲  at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
14:05:32.306 | ▲  > Build error occurred
14:05:32.309 | ▲  Error: Failed to collect page data for /register
14:05:32.309 | ▲    at /opt/buildhome/repo/node_modules/.pnpm/next@13.4.12_@babel+core@7.22.9_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/build/utils.js:1156:15
14:05:32.309 | ▲    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
```

**Comment by emrecoban** [comment]
@thorwebdev You are real hero! Everything is working perfectly! I've worked on this issue for **3 days**. You saved me and my project! If you come to Turkey one day, ping me please! :)

```javascript
import { cookies } from 'next/headers';
import { cache } from 'react';
import { createServerComponentClient } from '@supabase/auth-helpers-nextjs';

export const createServerSupabaseClient = cache(() => {
    const cookieStore = cookies()
    return createServerComponentClient({ cookies: () => cookieStore })
})

export async function getSession() {
    const supabase = createServerSupabaseClient()
    try {
        const {
            data: { session }
        } = await supabase.auth.getSession()
        return session
    } catch (error) {
        console.error('Error:', error)
        return null
    }
}
```

**Comment by EdgarDev17** [comment]
Hey people! I am developing a project and getting the same error so I tried some solutions you posted here, but now I am getting this error. "Error: TypeError: this.context.cookies is not a function"

```javascript
// app/propietario/dashboard/propiedades/[id]/page.tsx
// id es el segmento dinámico de la ruta ☝🏽

import {cookies} from 'next/headers'
import { cache } from 'react';
import {createServerComponentClient} from '@supabase/auth-helpers-nextjs'

export const createServerSupabaseClient = cache(() => {
    const cookieStore = cookies()
    return createServerComponentClient({ cookies: () => cookieStore })
})

export async function generateStaticParams() {

	const supabase = createServerSupabaseClient()

	const {data, error} = await supabase.from('propiedades').select()

	if (error) {
		throw error
	}

	return data.map((property) => ({
		id: property.id,
	}))
}

export default function Page({params}: {params: {id: string}}) {
	return <div>Esta es la propiedad: {params.id}</div>
}

```

**Comment by dbustosp** [comment]
@thorwebdev 
Thanks for the solution!! This worked like a charm! 
Below is my snippet: 

```
import { createServerComponentClient } from "@supabase/auth-helpers-nextjs";
import { Database } from "@/lib/types/schema";
import { cookies } from 'next/headers'
import { cache } from 'react';

export const createServerClient = cache(() => {
    const cookieStore = cookies()
    return createServerComponentClient<Database>({
        cookies: () => cookieStore
    })
})

const supabase = createServerClient()

export async function GET() {
    const supabase = createServerClient()

    const { data, error } = await supabase.from('').select('')

    return data
}
```

**Comment by IB1899** [comment]
I have the this issue when I try to access the **cookies()** from _next/headers_ in the edge runtime.
is there a reason why I can't access it there?

**Comment by arthureberledev** [comment]
> I have the this issue when I try to access the **cookies()** from _next/headers_ in the edge runtime. is there a reason why I can't access it there?

What solved it for me was using the `@edge-runtime/cookies` lib from vercel and doing the following:

```
import { RequestCookies } from "@edge-runtime/cookies";
import { createRouteHandlerClient } from "@supabase/auth-helpers-nextjs";

import type { Database } from "../types/db";

export function createEdgeHandlerClient(req: Request) {
  const cookies = new RequestCookies(req.headers) as any;
  return createRouteHandlerClient<Database>({ cookies: () => cookies });
}
```

**Comment by dacgray** [comment]
@IB1899 @arthureberledev 

If you're working in middleware, this works for me out of the box: `request.cookies.get('myCookie')?.value`

**Comment by a-ryang** [comment]
`export const dynamic = "force-dynamic";`
I looked at the following link and solved it
https://github.com/vercel/next.js/commit/2b497413087c56dedb3e04f77ea0deebb5d12d0b

**Comment by orenlande** [comment]
> > I have the this issue when I try to access the **cookies()** from _next/headers_ in the edge runtime. is there a reason why I can't access it there?
> 
> What solved it for me was using the `@edge-runtime/cookies` lib from vercel and doing the following:
> 
> ```
> import { RequestCookies } from "@edge-runtime/cookies";
> import { createRouteHandlerClient } from "@supabase/auth-helpers-nextjs";
> 
> import type { Database } from "../types/db";
> 
> export function createEdgeHandlerClient(req: Request) {
>   const cookies = new RequestCookies(req.headers) as any;
>   return createRouteHandlerClient<Database>({ cookies: () => cookies });
> }
> ```

It worked for me!

**Comment by wiesson** [comment]
> @thorwebdev You are real hero! Everything is working perfectly! I've worked on this issue for **3 days**. You saved me and my project! If you come to Turkey one day, ping me please! :)
> 
> ### Solution:
> ```js
> import { cookies } from 'next/headers';
> import { cache } from 'react';
> import { createServerComponentClient } from '@supabase/auth-helpers-nextjs';
> 
> export const createServerSupabaseClient = cache(() => {
>     const cookieStore = cookies()
>     return createServerComponentClient({ cookies: () => cookieStore })
> })
> 
> export async function getSession() {
>     const supabase = createServerSupabaseClient()
>     try {
>         const {
>             data: { session }
>         } = await supabase.auth.getSession()
>         return session
>     } catch (error) {
>         console.error('Error:', error)
>         return null
>     }
> }
> ```

Could live this function outside of `page` or `layout` files? It feels quite repetitive to write that all the time 😓

**Comment by balazsorban44** [comment]
Hi everyone, I just tested this using the original reproduction steps and it works on the latest `next@canary` version, so this issue can be closed. I recommend trying the same with your own code (please upgrade the related dependencies too), and if you are still seeing issues, please open a new one, with a minimal reproduction. Thanks!

**Comment by millertchris** [comment]
Here's a solution that worked for me.

I was struggling with the following error today while developing locally:
`Error: Invariant: cookies() expects to have requestAsyncStorage, none available.`

This was not an issue when I first started our project but it started overtime, so I figured it was something that I did to introduce this error. This forced me to go back and look at the following documentation.

https://supabase.com/docs/guides/auth/auth-helpers/nextjs#creating-a-supabase-client

In this documentation, it shows:
```
import { cookies } from 'next/headers'
import { createServerComponentClient } from '@supabase/auth-helpers-nextjs'

export const dynamic = 'force-dynamic'

export default async function Page() {
  const supabase = createServerComponentClient({ cookies })
  const { data } = await supabase.from('todos').select()
  return <pre>{JSON.stringify(data, null, 2)}</pre>
}
```

The important thing that I noticed here was that they are importing the `cookies` component before the `createServerComponentClient` — and in my case — I somehow reversed the two at some point in a few files.

I also noticed in several examples that people shared — this also seemed to be the case for them; here's an example:
```
import { createServerComponentClient } from "@supabase/auth-helpers-nextjs";
import { cookies, headers } from "next/headers";
```

So when I moved `import { cookies, headers } from "next/headers";` above `import { createServerComponentClient } from "@supabase/auth-helpers-nextjs";` it resolved the issue for me!

I found this thread when performing a Google search and hope that this helps someone else if they also find themselves facing this same issue.

Using `Next.js 13` and `@supabase/auth-helpers-nextjs v0.8.1`

@WilderDev @kelvinsekx @emrecoban @thorwebdev @dbustosp @a-ryang
